libs/langchain-community/src/chat_models/friendli.ts

```diff
@@ -0,0 +1,378 @@
+import { CallbackManagerForLLMRun } from "@langchain/core/callbacks/manager";
+import {
+  BaseChatModel,
+  BaseChatModelCallOptions,
+  type BaseChatModelParams,
+} from "@langchain/core/language_models/chat_models";
+import {
+  BaseMessage,
+  AIMessage,
+  ChatMessage,
+  HumanMessage,
+  SystemMessage,
+  HumanMessageChunk,
+  AIMessageChunk,
+  SystemMessageChunk,
+  ChatMessageChunk,
+} from "@langchain/core/messages";
+import {
+  ChatGeneration,
+  ChatGenerationChunk,
+  ChatResult,
+} from "@langchain/core/outputs";
+import { getEnvironmentVariable } from "@langchain/core/utils/env";
+import { convertEventStreamToIterableReadableDataStream } from "../utils/event_source_parse.js";
+
+/**
+ * Type representing the role of a message in the Friendli chat model.
+ */
+export type FriendliMessageRole = "system" | "assistant" | "user";
+
+interface FriendliMessage {
+  role: FriendliMessageRole;
+  content: string;
+}
+
+function messageToFriendliRole(message: BaseMessage): FriendliMessageRole {
+  const type = message._getType();
+  switch (type) {
+    case "ai":
+      return "assistant";
+    case "human":
+      return "user";
+    case "system":
+      return "system";
+    case "function":
+      throw new Error("Function messages not supported");
+    case "generic": {
+      if (!ChatMessage.isInstance(message)) {
+        throw new Error("Invalid generic chat message");
+      }
+      if (["system", "assistant", "user"].includes(message.role)) {
+        return message.role as FriendliMessageRole;
+      }
+      throw new Error(`Unknown message type: ${type}`);
+    }
+    default:
+      throw new Error(`Unknown message type: ${type}`);
+  }
+}
+
+function friendliResponseToChatMessage(message: FriendliMessage): BaseMessage {
+  switch (message.role) {
+    case "user":
+      return new HumanMessage(message.content ?? "");
+    case "assistant":
+      return new AIMessage(message.content ?? "");
+    case "system":
+      return new SystemMessage(message.content ?? "");
+    default:
+      return new ChatMessage(message.content ?? "", message.role ?? "unknown");
+  }
+}
+
+function _convertDeltaToMessageChunk(
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  delta: Record<string, any>
+) {
+  const { role } = delta;
```

Comment:
This is probably a bug in the Friendli API docs, but the example response shows that `role` isn't always returned in the streaming response. If this is true, then this will cause an error.

I'm looking at this documentation: https://docs.friendli.ai/openapi/create-chat-completions#responses

Reply:
I added defaultValue("assistant") for `role`.